File: backend/app/services/auth_service.py
from datetime import timedelta
from fastapi import HTTPException, status
from sqlalchemy.orm import Session
from app.core.config import settings
from app.core.security import create_access_token, get_password_hash, verify_password
from app.repositories import user_repository
from app.schemas.auth import LoginRequest, TokenResponse
from app.schemas.user import UserCreate


def _token_response(user) -> TokenResponse:
    token = create_access_token(
        {"sub": user.id},
        expires_delta=timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES),
    )
    return TokenResponse(access_token=token, token_type="bearer", user=user)


def register_user(db: Session, payload: UserCreate) -> TokenResponse:
    existing = user_repository.get_by_email(db, payload.email.lower())
    if existing:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Email is already registered")

    user = user_repository.create_user(
        db,
        full_name=payload.full_name,
        email=payload.email,
        password_hash=get_password_hash(payload.password),
        phone=payload.phone,
        address=payload.address,
    )
    return _token_response(user)


def login_user(db: Session, payload: LoginRequest) -> TokenResponse:
    user = user_repository.get_by_email(db, payload.email.lower())
    if not user or not verify_password(payload.password, user.password_hash):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid email or password")
    if not user.is_active:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="User account is disabled")
    return _token_response(user)

File: src/services/authService.ts
import { apiClient } from "./apiClient";
import type { AuthSession, LoginRequest, RegisterRequest } from "../types/auth";

interface BackendAuthResponse {
  access_token: string;
  token_type: "bearer";
  user: {
    id: number;
    full_name: string;
    email: string;
    role: "customer" | "admin";
  };
}

function mapAuth(response: BackendAuthResponse): AuthSession {
  return {
    token: response.access_token,
    user: {
      id: String(response.user.id),
      name: response.user.full_name,
      email: response.user.email,
      role: response.user.role,
    },
  };
}

export async function login(payload: LoginRequest): Promise<AuthSession> {
  const response = await apiClient.post<BackendAuthResponse>("/api/auth/login", payload);
  return mapAuth(response);
}

export async function register(payload: RegisterRequest): Promise<AuthSession> {
  const response = await apiClient.post<BackendAuthResponse>("/api/auth/register", {
    full_name: payload.name,
    email: payload.email,
    password: payload.password,
    phone: payload.phone,
  });
  return mapAuth(response);
}
